source/common/upstream/cluster_manager_impl.cc

```diff
@@ -431,6 +431,19 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   // We have a situation that clusters will be immediately active, such as static and primary
   // cluster. So we must have this prevention logic here.
   if (cluster_data != warming_clusters_.end()) {
+    // If there is no secret entity, currently supports only TLS Certificate and Validation
+    // Context, when it failed to extract them via SDS, it will fail to change cluster status from
+    // warming to active. In current implementation, there is no strategy to activate clusters
+    // which failed to initialize at once.
+    // TODO(shikugawa): To implement to be available by keeping warming after no-available secret
+    // entity behavior occurred. And remove
+    // `envoy.reloadable_features.cluster_keep_warming_no_secret_entity` runtime feature flag.
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.cluster_keep_warming_no_secret_entity") &&
+        !cluster.info()->transportSocketMatcher().resolve(nullptr).factory_.isReady()) {
+      ENVOY_LOG(warn, "Failed to activate {}", cluster.info()->name());
+      return;
```

Comment:
How does the cluster ever activate? I think this function never gets called again?

Reply:
The idea is keep the cluster warming unless it has the a valid secret entity since it can't be used in meaningful way, because the cluster will end up using NotReadySslSocket and return 503s always. That's what the TODO above is about (https://github.com/envoyproxy/envoy/issues/13952) and the long term fix https://github.com/envoyproxy/envoy/issues/13952

Reply:
But how can the cluster ever leave warming? It needs to get updated again by the management server?

Reply:
Right, until #13777 (that's the TODO in the previous comment but I pasted wrong)

Reply:
But stepping back, what does this change give us? If the cluster stays in warming and the server starts up, we will still get 503s? Or if a cluster gets stuck in warming even if the user sets a timeout, it seems like it would be extremely difficult to debug this?

What is the exact scenario we are trying to protect against? I assume the server came up and SDS is not ready during a CDS update? Because I don't think this woulds block initial server startup?

Reply:
Yes you still get 503s but it does affect the initialization sequence. See #11120 for more context.

Reply:
Can you summarize? There is a lot of text in that issue. Or better can you just beef up the comments and then I can read that? It would be nice to have very clear comments here on how this effects initialization flow and why this is better than what we had before given that it will be warming forever.

/wait

Reply:
I think there are two related issues. The first is the cluster is marked ready immediately. This means that if SDS doesn't return the certs, we still mark everything ready. For us, we look at this readiness to determine if we should start sending traffic to the pod, so now we will start sending traffic to a ""ready"" envoy which will always fail. Ideally, if the required secrets are not there, it would be in warming, so we do not send traffic. Once the secret is returned, it could be marked ready.

The other issue, which I am not sure if it is addressed here or another PR, is that occasionally we see that Envoy fails to actually send an SDS request to the control plane, so the cluster never actually gets ready. This is exacerbated by the previous issue, because not only will we never be able to handle (tls) traffic, we will also mark ourselves as ready.

Reply:
@mattklein123 to be clear, this is not a completely satisfactory fix for the underlying issu. That's why it is guarded in a runtime flag and have issue for long term fix (#13952), but this solves the issue in short term, it just mitigates the not ready issue.

Reply:
OK that's fine, but I don't really understand what the implications of this are and whether it's going to cause more confusing behavior that is also hard to debug. Can you add more comments and we can discuss? Do we need stats for this also? I just don't know what we are dealing with here honestly.

Reply:
IMO this won't cause more confusing behavior. While I agree that the current behavior is still confusing because the cluster with not ready ssl will fail when it gets traffic. The problem that this PR fix is that Envoy won't advertise itself is ready in that case.

The runtime flag is off by default and will be removed when we have long term fix. I don't think we need a stats for this given we have a warn level log and the flag is off by default.

Reply:
OK can you add more comments and we can go from there? I think we are going in circles as I'm unclear on what the old/new behavior is. :)

Reply:
@mattklein123 reworded the comments.

Reply:
> The other issue, which I am not sure if it is addressed here or another PR, is that occasionally we see that Envoy fails to actually send an SDS request to the control plane, so the cluster never actually gets ready. This is exacerbated by the previous issue, because not only will we never be able to handle (tls) traffic, we will also mark ourselves as ready.

@howardjohn I have seen such an issue as well. Was this root caused to a bug in envoy?